### ISSUE TITLE
HugrView::get_function_type makes no sense

### ISSUE THREAD

**Issue by acl-cqc** [CQCL/hugr#957]
It returns the PolyFuncType of a function, but the FunctionType of any other container e.g. DFG *plus zero binders* - what is going on here?

* Used in constants, where we `try_into` the PolyFuncType *back* to a FunctionType
* And in test code
* Probably better to make it either
    * return a monomorphic function type or none (i.e. reject any FuncDecl/Defn with binders), as per the use in const; or
    * make it return a PolyFuncType for FuncDecl/fn, only (and None for anything else). It could then be used in the builder for building Call and LoadFunction nodes (which has its own code in `build_traits.rs` atm)

At the root here is the question - can you put a FuncDefn in a Constant and then LoadConstant it? Currently the rules say that you *can* (this lets you get a closure of a recursive function, which you can't do by LoadConstant'ing a DFG). But after #947 one could just add the FuncDefn to the Hugr and use LoadFunction. So maybe / do we really want to keep the FuncDefn-in-Constant just so we can have a *nested Hugr* rather than a child node?

**Comment by acl-cqc** [comment]
On the latter: the thing is that it only works for monomorphic FuncDefns, which seems a bit crappy. There ought to be a way to apply type args, in which case, we have LoadFunction. So (@ss2165) is it really that important to allow a FuncDefn-rooted Hugr in a Const (a separate Hugr) as opposed to putting that FuncDefn as a subtree of the main Hugr?

Options? I can see about three...
* A wild idea would be to make a function *value* store both a FuncDefn-rooted Hugr and TypeArgs, I guess then we'd have two kinds of function Value (another storing DFG-rooted Hugr without TypeArgs)

* `get_poly_function_type` (as you suggest) that returns None for e.g. DFGs seems useful, e.g. there are a couple of methods in builder that do this; then `mono_fn_type` (in constant.rs) could be built calling both `get_poly_function_type` and `get_df_function_type` (if the former is None)

* Or, `mono_fn_type` could be moved into HugrView and `get_function_type` removed by inlining into `mono_fn_type`, I think that would work too

**Comment by acl-cqc** [comment]
> Why can't we add Type Args to LoadConstant? Now polymorphic functions inside a Const would work fine? I'm not entirely sure keeping the LoadConstant/LoadFunction separation would be worth it if you did that.

Possibly we should have this discussion on PR #947, but...

The problem is that `EdgeKind::Const` has a standard `Type`, which cannot express a polymorphic function. The alternative is an `EdgeKind::Function`, which has both a type and a list binders (`TypeParam`s) - possibly empty. So, I think your proposal says:
* `LoadConstant` has a list of typeargs, which may be empty
* If the list of typeargs is empty, the input is an `EdgeKind::Const` of a monomorphic (perhaps-function) type
* If the list of typeargs is non-empty, the "body" of the type must be a function type, and the input is an `EdgeKind::Function` (i.e. the LoadConstant will store the TypeParams as well as the TypeArgs - like `Call` and #947's LoadFunction, in the PolyFuncType).

Since you intend to `LoadConstant` a `FuncDefn`, we have to define a similar rule there, i.e.
* If the FuncDefn declares (a non-empty list of) TypeParams, its output is an `EdgeKind::Function`, otherwise its an `EdgeKind::Const`

And thus to work with that, `Call` then has to have the same behaviour as `LoadConstant`. Finally, `Const`...
* May contain a `DFG`-rooted Hugr --> has output an `EdgeKind::Const` (no `TypeParam`s)
* May contain a `FuncDefn`-rooted Hugr --> has output the same edgekind as the `FuncDefn`

To my mind, this has an awful lot of "node X does A if P else it does B" and, rather than us trying to automatically do "the right thing", it's better to make the Hugr-user be explicit about what they want. Thus, I propose:

* `FuncDefn` always has an `EdgeKind::Function` output, whose list of TypeParams may be empty
* `Call` always has an `EdgeKind::Function` input, and a list of TypeArgs (other inputs/outputs appropriate to the function)
* `LoadFunction` has an `EdgeKind::Function` input, and a list of TypeArgs (and a single `EdgeKind::Const` output, containing a mono fn type). Perhaps it should be called `LoadLambda` or something.
* `Const` may contain only (non-function values or) DFG-rooted Hugrs (I guess perhaps also CFG-rooted), and has a `EdgeKind::Const` output (i.e. perhaps a monomorphic function type). If we really want, we *could* allow a `Const` to contain a `FuncDefn`-rooted Hugr with no typeargs, but I think it's better to force use of:
* `FuncConst` contains a `FuncDefn`-rooted Hugr (always, no other kind of value), and has an `EdgeKind::Function` output (so can be used with `Call` or `LoadFunction` only) 
* `LoadConstant` always has an `EdgeKind::Const` input (i.e. which contains a type which is transferred directly onto the `EdgeKind::Value` output)

Once again I'll ping @ss2165 as to how much we *really* want to be able to have *both* nodes with function-bodies defined by their children in the hierarchy, *and* nodes with function-bodies defined by a separate `struct Hugr` contained in the node weight. But if we don't care too much about that, we just skip `FuncConst`.

|   | EdgeKind::Function | EdgeKind::Const |
|--| ---------- | ---------- |
| Input | `LoadFunction` (#947, produces thunk), `Call` | `LoadConstant` (produces thunk, or other value) |
| Output | `FuncDefn` (body defined by children), `FuncConst` (body defined by Hugr-value, not yot realized) | `Const` |

**Comment by acl-cqc** [comment]
Haha - yes we did consider that in #906 - indeed the PR started out that way!! But yes, we could change back to that fairly easily.

> I think `LoadFunction` is really `TypeApply`

yes, that is true, or at least, that is an excellent way of looking at it!

The other thing we haven't talked about at all (and we should any time we are thinking about FuncDefn's, really) is linking #882.

It might also be possible to simplify by *removing* DFG-rooted-Hugrs in Consts. Any such can easily be converted to a FuncDefn-rooted Hugr with no TypeParams in a FuncConst, and any LoadConstant successor-node into a LoadFunction with no TypeArgs. (Since edges from Consts go directly to LoadConstants, only, this is straightforward.) The only real requirement there is to come up with a name to put in the FuncDefn but this is of no importance (name could even become optional).

**Comment by ss2165** [comment]
Apologies for being late to this - kept waiting for an opportunity to discuss in person that didn't arise.

Just had a conversation with @doug-q and @mark-koch where I believe everyone left on the same page. I will attempt to summarise. (I don't have a strong opinion on `get_function_type` - I'm happy with any sensible refactor in to two methods for getting only polymorphic types and getting only monomorphic types). 

My first question was why allow hugr-valued constants if you can just `LoadFunction` - the answer is that constants can't have static edges coming in to them, they're self-contained (which we can make good use of for prepare-kernel-then-execute workflows).


In answer to the question I think we are in this scenario: 
> But if we don't care too much about that, we just skip FuncConst.
So I'm going to consider the rest of that proposal without `FuncConst`

The only two remaining queries I have on that proposal are: 

- I think we should enforce at `Const` (really `Value` now) construction time that any HUGR can only have DFG or CFG root. I don't see a case for any other root, given it must be monomorphic.
- In #947 `LoadFunction` outputs a `Value` (which has a nice symmetry to `LoadConst`) rather than a `Const` as proposed . As far as I can tell the only difference is where `LoadFunction` would be valid (since the only thing you can do with a `Const` output is `LoadConst` it), as proposed above it would be valid in any static scope (not just inside dataflow blocks). I can't see why this would be particularly useful so I'm inclined to go with `Value` output.

**Comment by acl-cqc** [comment]
Coming back to this but not sure whether I should reopen....

> I think we should enforce at Const (really Value now) construction time that any HUGR can only have DFG or CFG root. I don't see a case for any other root, given it must be monomorphic.

The one thing you can do in a monomorphic FuncDefn(-rooted Hugr) that you can't do in a DFG-rooted Hugr is - recurse. That might be enough to justify `Const`s containing the former.

> * In [feat: Add LoadFunction node #947](https://github.com/CQCL/hugr/pull/947) `LoadFunction` outputs a `Value` (which has a nice symmetry to `LoadConst`) rather than a `Const` as proposed . As far as I can tell the only difference is where `LoadFunction` would be valid (since the only thing you can do with a `Const` output is `LoadConst` it), as proposed above it would be valid in any static scope (not just inside dataflow blocks). I can't see why this would be particularly useful so I'm inclined to go with `Value` output.

sorry, **where do we propose that `LoadFunction` outputs a `Const`**? Yes, one could have such a conversion (from a static Function edge, which could be polymorphic, to a static Const edge, which must be monomorphic), but the node to do that is therefore also *static* (does not execute at runtime), which breaks the so-far-implicit constraint that static/Module-level ops do not have inputs, and it seems a bit of a corner-case. What would you do with the `Const` output - wire it to a `LoadConst`? In which case surely keep `LoadFunction` doing both steps (`Function` right through to `Value`) in one go....

**Comment by acl-cqc** [comment]
Yes I don't object to that `TypeApply`, commoning up is an advantage, OTOH I think it does make static nodes more complicated, so I'm fairly neutral.

> You can do this in a DFG rooted hugr by including a scoped definition FuncDefn.

Right - in that case, surely the *only* `Const` Hugr should be DFG-rooted, then? You can stick a CFG, TailLoop or Conditional inside one of those, too...
